doc/zypp-NameReqPrv.1
'\" t
.\"     Title: zypp-namereqprv
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 06/12/2015
.\"    Manual: LIBZYPP
.\"    Source: SUSE Linux
.\"  Language: English
.\"
.TH "ZYPP\-NAMEREQPRV" "1" "06/12/2015" "SUSE Linux" "LIBZYPP"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
zypp-NameReqPrv \- Investigate packages and dependencies in solver test\-cases
.SH "SYNOPSIS"
.sp
\fBzypp\-NameReqPrv\fR [\fB\-\-root\fR \fIDIR\fR] [[\fIOPTIONS\fR] \fINAME\fR\&...]\&...
.SH "DESCRIPTION"
.sp
Load all enabled repositories (without refresh) and search for occurrences of regular expression \fINAME\fR in package names or dependencies\&. In case \fIDIR\fR denotes a directory containing a solver test\-case, repositories and settings provided by the test\-case are loaded instead\&.
.sp
A solver test\-case is an abstract of repositories, packages, hardware and dependency resolution related settings which can be created by using zyppers \fB\-\-debug\-solver\fR option\&. Attached to a bug report the test\-case helps investigating the reported behavior\&. Options and output of this command are tailored to this use\-case and may change as needed\&.
.sp
Reported matches show the packages \fIinternal id\fR, \fIname\fR, \fIversion\fR and \fIarchitecture\fR, \fIpriority\fR and name of the providing \fIrepository\fR, \fIvendor\fR, \fIbuildtime\fR and the locations matching \fINAME\fR\&.
.sp
.if n \{\
.RS 4
.\}
.nf
2431  zypper\-1\&.9\&.16\-22\&.2\&.x86_64  (99)@System  openSUSE  1400499579
                                              nam: zypper
40947  zypper\-1\&.9\&.3\-1\&.1\&.x86_64    (99)repo\-oss (13\&.1)  openSUSE     1383049437
                                                       nam: zypper
56140  zypper\-1\&.9\&.16\-22\&.2\&.x86_64  (99)repo\-oss\-update (13\&.1)  openSUSE     1400499579
                                                              nam: zypper
.fi
.if n \{\
.RE
.\}
.SH "OPTIONS"
.PP
\fB\-\-root\fR \fIDIR\fR
.RS 4
Load repos from the system located below
\fIDIR\fR\&. If
\fIDIR\fR
denotes a solver test\-case, the test\-case is loaded\&.
.RE
.PP
\fB\-\-installed\fR
.RS 4
Process installed packages only\&.
.RE
.PP
\fB\-i\fR/\fB\-I\fR
.RS 4
Turn on/off case insensitive search (default on)
.RE
.PP
\fB\-n\fR/\fB\-N\fR
.RS 4
Turn on/off looking for names (default on)
.RE
.PP
\fB\-p\fR/\fB\-P\fR
.RS 4
Turn on/off looking for provides (default off)
.RE
.PP
\fB\-r\fR/\fB\-R\fR
.RS 4
Turn on/off looking for requires (default off)
.RE
.PP
\fB\-c\fR/\fB\-C\fR
.RS 4
Turn on/off looking for conflicts (default off)
.RE
.PP
\fB\-o\fR/\fB\-O\fR
.RS 4
Turn on/off looking for obsoletes (default off)
.RE
.PP
\fB\-m\fR/\fB\-M\fR
.RS 4
Turn on/off looking for recommends (default off)
.RE
.PP
\fB\-s\fR/\fB\-S\fR
.RS 4
Turn on/off looking for supplements (default off)
.RE
.PP
\fB\-a\fR
.RS 4
Short for
\fB\-n \-p \-r\fR
.RE
.PP
\fB\-A\fR
.RS 4
Short for
\fB\-n \-P \-R\fR
.RE
.PP
\fB\-D\fR \fIPKG\fR
.RS 4
Dump dependencies of package
\fIPKG\fR
(exact name)\&.
.RE
.SH "AUTHORS"
.sp
Michael Andres <ma@suse\&.de>
.SH "SEE ALSO"
.sp
zypper(8)


po/ChangeLog
2006-01-27  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.14.3.
	* boldquot.sed: New file, from gettext-0.14.3.
	* en@boldquot.header: New file, from gettext-0.14.3.
	* en@quot.header: New file, from gettext-0.14.3.
	* insert-header.sin: New file, from gettext-0.14.3.
	* quot.sed: New file, from gettext-0.14.3.
	* remove-potcdate.sin: New file, from gettext-0.14.3.
	* Rules-quot: New file, from gettext-0.14.3.
	* POTFILES.in: New file.



tools/zypp-NameReqPrv.cc
#define INCLUDE_TESTSETUP_WITHOUT_BOOST
#include "zypp/../tests/lib/TestSetup.h"
#undef  INCLUDE_TESTSETUP_WITHOUT_BOOST

#include <algorithm>
#include <zypp/PoolQuery.h>
#include <zypp/ResObjects.h>

static std::string appname( "NameReqPrv" );

#define message cout
using std::flush;

int errexit( const std::string & msg_r = std::string(), int exit_r = 100 )
{
  if ( ! msg_r.empty() )
  {
    cerr << endl << msg_r << endl << endl;
  }
  return exit_r;
}

int usage( const std::string & msg_r = std::string(), int exit_r = 100 )
{
  if ( ! msg_r.empty() )
  {
    cerr << endl << msg_r << endl << endl;
  }
  cerr << "Usage: " << appname << " [--root ROOTDIR] [OPTIONS] NAME... [[OPTIONS] NAME...]..." << endl;
  cerr << "  Load all enabled repositories (no refresh) and search for" << endl;
  cerr << "  occurrences of NAME (regex) in package names or dependencies" << endl;
  cerr << "  --root   Load repos from the system located below ROOTDIR. If ROOTDIR" << endl;
  cerr << "           denotes a sover testcase, the testcase is loaded." << endl;
  cerr << "  --installed Process installed packages only." << endl;
  cerr << "  -i/-I    turn on/off case insensitive search (default on)" << endl;
  cerr << "  -n/-N    turn on/off looking for names       (default on)" << endl;
  cerr << "  -p/-P    turn on/off looking for provides    (default off)" << endl;
  cerr << "  -r/-R    turn on/off looking for requires    (default off)" << endl;
  cerr << "  -c/-C    turn on/off looking for conflicts   (default off)" << endl;
  cerr << "  -o/-O    turn on/off looking for obsoletes   (default off)" << endl;
  cerr << "  -m/-M    turn on/off looking for recommends  (default off)" << endl;
  cerr << "  -s/-S    turn on/off looking for supplements (default off)" << endl;
  cerr << "  -a       short for -n -p -r" << endl;
  cerr << "  -A       short for -n -P -R" << endl;
  cerr << "  -D <pkg> dump dependencies of <pkg>" << endl;
  cerr << "" << endl;
  return exit_r;
}

void tableOut( const std::string & s1 = std::string(),
               const std::string & s2 = std::string(),
               const std::string & s3 = std::string(),
               const std::string & s4 = std::string(),
               const std::string & s5 = std::string() )
{
  message << "  ";
#define TABEL(N) static unsigned w##N = 0; if ( ! s##N.empty() ) w##N = std::max( w##N, unsigned(s##N.size()) ); message << str::form( " %-*s ", w##N, s##N.c_str() )
#define TABER(N) static unsigned w##N = 0; if ( ! s##N.empty() ) w##N = std::max( w##N, unsigned(s##N.size()) ); message << str::form( " %*s ", w##N, s##N.c_str() )
  TABER( 1 ); TABEL( 2 ); TABEL( 3 ); TABEL( 4 ); TABEL( 5 );
#undef TABEL
  message << endl;
}


///////////////////////////////////////////////////////////////////

void dDump( const std::string & spec_r )
{
  message << "DUMP " << spec_r << " {";

  sat::WhatProvides q( Capability::guessPackageSpec( spec_r ) );
  if ( q.empty() )
  {
    message << "}" << endl;
    return;
  }

  for ( const auto & el : q )
  {
    message << endl << "==============================" << endl << dump(el);
    if ( isKind<Pattern>(el) )
    {
      message << endl << "CONTENT: " << make<Pattern>(el)->contents();
    }
  }
  message << endl << "}" << endl;
}

/******************************************************************
**
**      FUNCTION NAME : main
**      FUNCTION TYPE : int
*/
int main( int argc, char * argv[] )
{
  INT << "===[START]==========================================" << endl;
  appname = Pathname::basename( argv[0] );
  --argc,++argv;

  if ( ! argc )
  {
    return usage();
  }

  ///////////////////////////////////////////////////////////////////

  ZConfig::instance();
  Pathname sysRoot("/");
  sat::Pool satpool( sat::Pool::instance() );

  if ( argc && (*argv) == std::string("--root") )
  {
    --argc,++argv;
    if ( ! argc )
      return errexit("--root requires an argument.");

    if ( ! PathInfo( *argv ).isDir() )
      return errexit("--root requires a directory.");

    sysRoot = *argv;
    --argc,++argv;
  }

  bool onlyInstalled( false );
  if ( argc && (*argv) == std::string("--installed") )
  {
    --argc,++argv;
    onlyInstalled = true;
  }

  if ( TestSetup::isTestcase( sysRoot ) )
  {
    message << str::form( "*** Load Testcase from '%s'", sysRoot.c_str() ) << endl;
    TestSetup test;
    test.loadTestcaseRepos( sysRoot );
    dumpRange( message, satpool.reposBegin(), satpool.reposEnd() ) << endl;
  }
  else if ( TestSetup::isTestSetup( sysRoot ) )
  {
    message << str::form( "*** Load TestSetup from '%s'", sysRoot.c_str() ) << endl;
    TestSetup test( sysRoot, Arch_x86_64 );
    test.loadRepos();
    dumpRange( message, satpool.reposBegin(), satpool.reposEnd() ) << endl;
  }
  else
  {
    // a system
    message << str::form( "*** Load system at '%s'", sysRoot.c_str() ) << endl;
    if ( true )
    {
      message << "*** load target '" << Repository::systemRepoAlias() << "'\t" << endl;
      getZYpp()->initializeTarget( sysRoot );
      getZYpp()->target()->load();
      message << satpool.systemRepo() << endl;
    }

    if ( !onlyInstalled )
    {
      RepoManager repoManager( sysRoot );
      RepoInfoList repos = repoManager.knownRepositories();
      for_( it, repos.begin(), repos.end() )
      {
        RepoInfo & nrepo( *it );

        if ( ! nrepo.enabled() )
          continue;

        if ( ! repoManager.isCached( nrepo ) )
        {
          message << str::form( "*** omit uncached repo '%s' (do 'zypper refresh')", nrepo.name().c_str() ) << endl;
          continue;
        }

        message << str::form( "*** load repo '%s'\t", nrepo.name().c_str() ) << flush;
        try
        {
          repoManager.loadFromCache( nrepo );
          message << satpool.reposFind( nrepo.alias() ) << endl;
        }
        catch ( const Exception & exp )
        {
          message << exp.asString() + "\n" + exp.historyAsString() << endl;
          message << str::form( "*** omit broken repo '%s' (do 'zypper refresh')", nrepo.name().c_str() ) << endl;
          continue;
        }
      }
    }
  }

  ///////////////////////////////////////////////////////////////////

  bool ignorecase	( true );
  bool names		( true );
  bool provides		( false );
  bool requires		( false );
  bool conflicts	( false );
  bool obsoletes	( false );
  bool recommends	( false );
  bool supplements	( false );

  for ( ; argc; --argc,++argv )
  {
    if ( (*argv)[0] == '-' )
    {
      switch ( (*argv)[1] )
      {
        case 'a': names =	true, 	requires = provides =	true;	break;
        case 'A': names =	true, 	requires = provides =	false;	break;
	case 'D':
	  if ( argc > 1 )
	  {
	    --argc,++argv;
	    dDump( *argv );
	  }
	  else
	    return errexit("-D <pkgspec> requires an argument.");
	  break;
        case 'i': ignorecase =	true;	break;
        case 'I': ignorecase =	false;	break;
        case 'n': names =	true;	break;
        case 'N': names =	false;	break;
        case 'r': requires =	true;	break;
        case 'R': requires =	false;	break;
        case 'p': provides =	true;	break;
        case 'P': provides =	false;	break;
        case 'c': conflicts =	true;	break;
        case 'C': conflicts =	false;	break;
        case 'o': obsoletes =	true;	break;
        case 'O': obsoletes =	false;	break;
        case 'm': recommends =	true;	break;
        case 'M': recommends =	false;	break;
        case 's': supplements =	true;	break;
        case 'S': supplements =	false;	break;
      }
      continue;
    }

    PoolQuery q;
    if ( onlyInstalled )
      q.setInstalledOnly();
    std::string qstr( *argv );

    if ( *argv == ResKind::product )
    {
      q.addKind( ResKind::product );
    }
    else if ( *argv == ResKind::patch )
    {
      q.addKind( ResKind::patch );
    }
    else if ( *argv == ResKind::pattern )
    {
      q.addKind( ResKind::pattern );
    }
    else
    {
      sat::Solvable::SplitIdent ident( qstr );
      if ( ident.kind() != ResKind::package )
      {
	q.addKind( ident.kind() );
	q.addString( ident.name().asString() );
      }
      else
	q.addString( qstr );

      q.setMatchRegex();
      q.setCaseSensitive( ! ignorecase );

      if ( names )
	q.addAttribute( sat::SolvAttr::name );
      if ( provides )
	q.addDependency( sat::SolvAttr::provides );
      if ( requires )
	q.addDependency( sat::SolvAttr::requires );
      if ( conflicts )
	q.addDependency( sat::SolvAttr::conflicts );
      if ( obsoletes )
	q.addDependency( sat::SolvAttr::obsoletes );
      if ( recommends )
	q.addDependency( sat::SolvAttr::recommends );
      if ( supplements )
	q.addDependency( sat::SolvAttr::supplements );
    }

    message << *argv << " [" << (ignorecase?'i':'_') << (names?'n':'_') << (requires?'r':'_') << (provides?'p':'_')
    << (conflicts?'c':'_') << (obsoletes?'o':'_') << (recommends?'m':'_') << (supplements?'s':'_') << "] {" << endl;

    for_( it, q.begin(), q.end() )
    {
      tableOut( str::numstring( it->id() ), it->asString(),
		str::form( "(%d)%s", it->repository().info().priority(), it->repository().name().c_str() ),
		str::numstring( PoolItem(*it)->buildtime() ) );
      tableOut( "", "",
		it->vendor().asString() );
      if ( ! it.matchesEmpty() )
      {
	for_( match, it.matchesBegin(), it.matchesEnd() )
	{
	  tableOut( "", "", match->inSolvAttr().asString().substr( 9, 3 )+": " +match->asString() );
	}
      }
    }

    message << "}" << endl;
  }

  INT << "===[END]============================================" << endl << endl;
  return 0;
}
